calendarのプログラム(JavaScript)版を実装しました

yumiya-g opened the pull request.

yumiya-g committed:
Add: 実行ファイルを作成しました

- `process.argv`プロパティでコマンドライン引数を取得できることを確認

yumiya-g committed:
Add: パッケージminimistをインストールしました

yumiya-g committed:
Add: パッケージDAY.JSをインストールしました

yumiya-g committed:
Fix: カレンダー出力を調整しました

yumiya-g committed:
Fix: 関数の処理の見直しと変数名の変更を行いました

- 調整後、ESLintとPrettierを実行

yumiya-g committed:
Add: Shebangの追加と変数todayの位置を修正しました

yuuu commented on 02.calendar/cal.js:
```diff
@@ -0,0 +1,55 @@
+const argv = minimist(process.argv.slice(2));
+const yearMonthValue = [argv.y, argv.m];
+const today = new Date();
+const year = yearMonthValue[0] || dayjs(today).$y;
```
ここですが `dayjs(today).year()` ではなく `$y` としているのは何か理由がありますか？

yumiya-g replied:
`dayjs(today)`でオブジェクトを取得して、`$y`キーから値を取得しておりました。
`year()`のほうがわかりやすいと感じましたので、関連する箇所を修正します。
https://day.js.org/docs/en/get-set/year

```javascript
// dayjs(today)
// #=>
M {
  '$L': 'ja',
  '$d': 2024-07-08T13:11:36.407Z,
  '$y': 2024,
  '$M': 6,
  '$D': 8,
  '$W': 1,
  '$H': 22,
  '$m': 11,
  '$s': 36,
  '$ms': 407,
  '$x': {},
  '$isDayjsObject': true
}
```

yuuu commented on 02.calendar/cal.js:
```diff
@@ -0,0 +1,55 @@
+
+const addSpaceLessThanTen = (days) => {
+  const daysValue = days.map((day) => {
+    return day < 10 ? ` ${day}` : `${day}`;
```
ここは `padStart` というメソッドを使うともっとシンプルに書くことができますね
https://developer.mozilla.org/ja/docs/Web/JavaScript/Reference/Global_Objects/String/padStart

yumiya-g replied:
テンプレートリテラルで空白を入れていましたが、`padStart`を使うとより明示的に文字列を調整したことがわかりやすくなると思いました。こちらも修正いたします。

yuuu commented on 02.calendar/cal.js:
```diff
@@ -0,0 +1,55 @@
+const argv = minimist(process.argv.slice(2));
+const yearMonthValue = [argv.y, argv.m];
+const today = new Date();
+const year = yearMonthValue[0] || dayjs(today).$y;
```
このあたりの変数が実質グローバル変数のような使われ方をしています。極力引数で受け渡すようにできないか検討をお願いします。

yumiya-g replied:
変数を減らして、カレンダーを出力する関数に「年」「月」を引数として渡す処理に変更を行いました。

yuuu commented on 02.calendar/cal.js:
```diff
@@ -0,0 +1,55 @@
+const month = yearMonthValue[1] || dayjs(today).$M + 1;
+const firstDay = dayjs(new Date(year, month, 1)).$D;
+const lastDay = dayjs(new Date(year, month, 0)).$D;
+const firstDayOfWeek = dayjs(`"${year}-${month}-${firstDay}"`).$W;
```
ここは `dayjs([year, month - 1, firstDay])` だと都合が悪いでしょうか？
https://day.js.org/docs/en/parse/array

yumiya-g replied:
`dayjs`に配列で「年」「月」「日」を渡す処理に変更しました🙏
※ただ`month - 1`のように`month`から「-1」すると、1月ずれるため`dayjs([year, month])`として対応しました。

yuuu commented on 02.calendar/cal.js:
```diff
@@ -0,0 +1,55 @@
+
+const createFirstDayToLastday = (lastDay) => {
+  const monthOfDays = [];
+  for (let i = 1; i <= lastDay; i++) {
```
複数のメソッドで何度も日付での繰り返し処理を記述していますが、この方が見やすいでしょうか？1回の繰り返し処理で「土曜日の改行」「日付のpadding」をした方がコードもシンプルになる、といったことはないでしょうか？

yumiya-g replied:
当初1つのメソッドにまとめた際に行数が多くなり見通しがわるいと感じたので、「土曜日の改行」「日付のpadding」「初日の曜日調整」の3つのタスクをメソッドに分解しました😓

頂いたアドバイスから、処理を見直して再度1つのメソッドにまとめてみました。

yumiya-g committed:
Fix: Day.jsからゲッターを使って日時情報を取得する処理に変更しました

- https://day.js.org/docs/en/get-set/get-set

yumiya-g committed:
Fix: `Days.js`の日付を取得する処理を、配列に変更しました

- https://day.js.org/docs/en/parse/array

yumiya-g committed:
Fix: 日付の表示調整の処理を変更しました

- `padStart`,`padEnd`でpaddingを調整

yumiya-g committed:
Fix: 変数の設定とカレンダーの表示処理を調整しました

- 変数をできるだけ減らし、「年」「月」を引数にして渡す処理に変更
- 指定月の日付を取得する関数名を`createAllDatesInMonth`に変更
- カレンダーを表示する処理を`outputCalendar`関数にまとめた

yuuu commented on 02.calendar/cal.js:
```diff
@@ -0,0 +1,38 @@
+    dayjs([year, month]).endOf("month").date(),
+  );
+  const adjustAllDates = allDays.map((day) => {
+    if (day === 1) {
```
1日にだけ必要な処理であれば繰り返し処理の前に行うことで、if文を1つ減らすことができますね。

yumiya-g replied:
`map`内に1日の処理をいれる必要はなかったので、修正します😓

yuuu commented on 02.calendar/cal.js:
```diff
@@ -0,0 +1,38 @@
+    } else if (1 < day && day < 10) {
+      day = day.toString().padStart(2, " ");
+    }
+    return dayjs([year, month, day]).day() === 6 ? day + "\n" : day + " ";
```
三項演算子のどちら側の処理を通っても `day +` を行っているので、 `day +` の右辺のみを三項演算子で切り替えるようにするとコードの重複をさらに減らせますね。

yumiya-g replied:
`day`のあとに三項演算子で改行または空白が入るよう調整します。

yumiya-g committed:
Fix: 1日の位置調整と、日付間の余白・改行の追加処理を修正しました

- 1日の表示開始位置調整を、`map`メソッド外に配置
- 余白または改行を追加する三項演算子を修正

yuuu commented on 02.calendar/cal.js:
```diff
@@ -23,14 +23,16 @@ const outputCalendar = (year, month) => {
   const allDays = createAllDatesInMonth(
     dayjs([year, month]).endOf("month").date(),
   );
+  if (allDays[0] === 1) {
```
mapの外側に移動させたことで、このif文は不要になったのではないでしょうか？

yumiya-g replied:
ご指摘のとおり`map`の外に配置したので不要な`if`文でした😓

yuuu commented on 02.calendar/cal.js:
```diff
@@ -23,14 +23,16 @@ const outputCalendar = (year, month) => {
-      const spaces = dayjs([year, month]).day() * 3 + 2;
-      day = day.toString().padStart(spaces, " ");
-    } else if (1 < day && day < 10) {
+    if (1 < day && day < 10) {
```
padStartがいい感じにしてくれるので、このif文も不要な認識です 🙏

yumiya-g replied:
ここも`padStart`で2桁表示にしていたので、2以上10未満の条件分岐も不要でした🙏

yumiya-g committed:
Fix: 不要な`if`文を削除しました
